### PR TITLE
Fix Issue 16724 - RandomCover.popFront is a no-op for the first call

### PR DESCRIPTION
**Pull request by RazvanN7**
Solved the bug. Now randomCover.popFront and length work correctly

### CODE REVIEW

File: std/random.d [hunk 1]
```diff
@@ -2198,39 +2207,38 @@ struct RandomCover(Range, UniformRNG = void)
     {
         @property size_t length()
         {
-            if (_alreadyChosen == 0)
-            {
-                return _input.length;
-            }
+            if (_isEmpty)
+                return 0;
             else
```
Comment (andralex) [inline std/random.d]: this else became redundant

File: std/random.d [hunk 2]
```diff
@@ -2198,39 +2207,38 @@ struct RandomCover(Range, UniformRNG = void)
...
 
     @property auto ref front()
     {
-        if (_alreadyChosen == 0)
-        {
-            popFront();
-        }
+        import std.exception : enforce;
+
+        enforce(!_isEmpty, "Cannot use front on an empty range");
```
Comment (andralex) [inline std/random.d]: let's use assert here (long discussion)

File: std/random.d [hunk 3]
```diff
@@ -2198,39 +2207,38 @@ struct RandomCover(Range, UniformRNG = void)
...
         return _input[_current];
     }
 
     void popFront()
     {
-        if (_alreadyChosen >= _input.length)
+        if (_isEmpty)
+            return;
```
Comment (wilzbach) [inline std/random.d]: FWIW this is undefined behavior - the range API explicitely forbids this, so the check is unnecessary:

> r.popFront can be called if and only if evaluating r.empty has, or would have, equaled false.

http://dlang.org/phobos/std_range_primitives.html

File: std/random.d [hunk 4]
```diff
@@ -2198,39 +2207,37 @@ struct RandomCover(Range, UniformRNG = void)
...
-            if (_alreadyChosen == 0)
-            {
-                return _input.length;
-            }
-            else
-            {
-                return (1 + _input.length) - _alreadyChosen;
-            }
+            if (_isEmpty)
+                return 0;
```
Comment (andralex) [inline std/random.d]: No need for this, right?

Comment (RazvanN7) [inline std/random.d]: True. It was a leftover from past iterations. Deleted

File: std/random.d [hunk 5]
```diff
@@ -2198,39 +2207,37 @@ struct RandomCover(Range, UniformRNG = void)
...
         }
     }
 
     @property auto ref front()
     {
-        if (_alreadyChosen == 0)
-        {
-            popFront();
-        }
+        import std.exception : enforce;
```
Comment (RazvanN7) [inline std/random.d]: Done

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by andralex** [comment]
@RazvanN7 thx!

**Comment by WebDrake** [comment]
Sorry to have missed this, but I think the fix is itself broken, because it strips out the lazy `front` of `randomCover`, and because I think it may introduce some other issues in its own right.

First, random algorithms (like random cover) ideally need to have a truly lazy initial `front` value, otherwise you can run into trivial errors like:

```D
auto arr = [1, 2, 3, 4, 5];
auto cover = arr.randomCover;
cover.writeln;  // three 'different' covers
cover.writeln;  // but each of them with
cover.writeln;  // the same first value
```

... so the check on already-chosen elements in `front` wasn't there by accident (although it may have needed further work).

Second, I just tried running the above right now and got a never-ending stream of `5, 5, 5, 5, 5, ...` out of my computer before I killed it.  So I'm not sure that this new `randomCover` is working right at all :-(

**Comment by WebDrake** [comment]
(... BTW, not to nitpick but ... a commit message whose only content is `Nitpicks` ??!!)

**Comment by WebDrake** [comment]
> Second, I just tried running the above right now and got a never-ending stream of `5, 5, 5, 5, 5, ...` out of my computer before I killed it. So I'm not sure that this new `randomCover` is working right at all :-(

Just tried reverting the patch, and the never-ending stream still happens if the output of `randomCover` is passed through `writeln`.  So, this patch isn't responsible.

The point about lazy initial `front` still stands, though.

I'll look into both issues over the weekend if I can.

**Pull request merged**